### PR TITLE
performance improvements

### LINKED ISSUE #2063: Numpy Performance Improvements
I recently came across two small numpy tricks for better performance. I think we use similar constructs in a lot of places in the code.

**Faster norm2**
Given 2 vectors we tend to use `np.sum(a * b)` in most places. I found that `np.einsum('i,i', a, b)` is about twice as fast on a recent numpy version (1.14.5, conda-forge). For 2 random vectors with shape (3, )

`np.sum(a,b)`: 3.3 micro-seconds
`np.einsum('i,i', a, b)`: 1.5 micro-seconds

**Faster Copying of Arrays**
A search shows that we use `a.copy()` around 180 times in the code. Often we have already allocated an array with the appropriate dimensions and would like to ideally copy directly into that skipping the reallocation. This happens often when we update `TimeStep` objects.

To illustrate take the following simplified example.

```python
# preallocate
a = np.empty(b.shape)
print(id(a))
# copy
a = b.copy()
print(id(a))
```

Here we preallocate a and copy b into it. In the second line python creates a new object that overwrites `a` and therefore reallocates. This can be seen by the changing id. This code therefore does **2** allocations. Instead look at

```python
a = np.empty(b.shape)
print(id(a))
np.copyto(a, b)
print(id(a))
```
The id of `a` did not change and no extra allocation occured.

### PR DESCRIPTION
**Pull request by shaivimalik**
Fixes #2063

Changes made in this Pull Request:
 - improves performance as discussed in Numpy Performance Improvements [#2063](https://github.com/MDAnalysis/mdanalysis/issues/2063)

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

----
:books: Documentation preview :books:: https://readthedocs-preview--4089.org.readthedocs.build/en/4089/

### CODE REVIEW

**Comment by shaivimalik** [comment]
@richardjgowers @orbeckst Hello! Can anybody review my PR?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Seems reasonable! Would you be able to see if this makes a change to the benchmarks? 

See: https://github.com/MDAnalysis/mdanalysis/blob/develop/benchmarks/README.rst

**Comment by shaivimalik** [comment]
@hmacdope Yes. I will get back to you with the result.

**Comment by shaivimalik** [comment]
@hmacdope I ran the benchmarks 3 times and got the following results:
Screenshot 2023-03-27 at 12 27 06 AM
Screenshot 2023-03-26 at 10 32 19 PM
Screenshot 2023-03-27 at 5 28 30 AM

**Comment by shaivimalik** [comment]
Hi, @hmacdope I made changes to more files where using np.einsum in place of np.sum seemed logical on a different branch and got the following benchmark results:
Screenshot 2023-03-27 at 10 57 50 PM
Screenshot 2023-03-28 at 12 37 24 AM
If these results are up to the mark, I will make these changes to the current PR branch.

**Comment by hmacdope** [comment]
The decrease in performance in some cases is unexpected. Did you have stuff running in the background at the same time  @g2707 ?

**Comment by shaivimalik** [comment]
@hmacdope No, I did not start any new applications during the benchmark.

**Comment by hmacdope** [comment]
@g2707 I can try on my computer tonight

**Comment by shaivimalik** [comment]
@hmacdope Should I add the new file changes before tonight?

**Comment by hmacdope** [comment]
Yep!

**Comment by shaivimalik** [comment]
Hi! @hmacdope Did you get a chance to run the benchmarks?

**Comment by hmacdope** [comment]
@g2707 yes I did thanks! Performance improvements look good!  2x runs shown below!

Screenshot from 2023-03-30 10-37-09
Screenshot from 2023-03-30 11-41-02

File: package/MDAnalysis/core/topologyattrs.py [hunk 1]
```diff
@@ -3305,4 +3304,4 @@ class CMaps(_Connection):
     attrname = 'cmaps'
     singular = 'cmaps'
     transplants = defaultdict(list)
-    _n_atoms = 5
+    _n_atoms = 5
```
Comment (hmacdope) [inline package/MDAnalysis/core/topologyattrs.py]: Change this back to the way it was.

File: package/MDAnalysis/lib/transformations.py [hunk 2]
```diff
@@ -1827,4 +1825,4 @@ def rotaxis(a, b):
 _import_module('_transformations')
 
 # Documentation in HTML format can be generated with Epydoc
-__docformat__ = "restructuredtext en"
+__docformat__ = "restructuredtext en"
```
Comment (hmacdope) [inline package/MDAnalysis/lib/transformations.py]: Same here

File: package/MDAnalysis/analysis/bat.py [hunk 3]
```diff
@@ -563,4 +564,4 @@ def Cartesian(self, bat_frame):
     @property
     def atoms(self):
         """The atomgroup for which BAT are computed (read-only property)"""
-        return self._ag
+        return self._ag
```
Comment (hmacdope) [inline package/MDAnalysis/analysis/bat.py]: Same here

**Review by hmacdope — CHANGES_REQUESTED** [review]
You will need a `CHANGELOG` entry (newest first), to add yourself to `AUTHORS` and as I think you are a new contributor we ask you to introduce yourself on the developer mailing list. ! Other than that looks good and all tests pass, just a few nitpicks.

File: package/MDAnalysis/analysis/bat.py (hunk 3 above)
Comment (shaivimalik) [reply package/MDAnalysis/analysis/bat.py]: Is it alright now?

File: package/AUTHORS [hunk 4]
```diff
@@ -203,6 +203,7 @@ Chronological list of authors
   - Patricio Barletta
   - Mikhail Glagolev
 2023
+  - Shaivi Malik
```
Comment (hmacdope) [inline package/AUTHORS]: Authors is last first.

**Review by hmacdope — CHANGES_REQUESTED** [review]

File: package/AUTHORS (hunk 4 above)
Comment (shaivimalik) [reply package/AUTHORS]: Done!

File: package/MDAnalysis/analysis/bat.py (hunk 3 above)
Comment (hmacdope) [reply package/MDAnalysis/analysis/bat.py]: No still needs to be changed.

Comment (shaivimalik) [reply package/MDAnalysis/analysis/bat.py]: should be fine now.

**Review by hmacdope — APPROVED** [review]
Looks good to me! Introduce yourself on the developer mailing list and then I can merge. :)

https://groups.google.com/g/mdnalysis-devel

**Comment by shaivimalik** [comment]
@hmacdope Done!

**Pull request merged**